### PR TITLE
Fix `auto_unmount` to work without `allow_other`

### PR DESCRIPTION
**Pull request by matthiasgoergens**
In
https://github.com/libfuse/libfuse/blob/77d662459a0fcdf358d515477d33795837e859d5/util/fusermount.c#L1219 `open` is executed as root which does not have access to the mount point if `allow_other` was not used and the real user id is not 0. Since `allow_other` usually cannot be specified by unprivileged users, `auto_unmount` has no effect for unprivileged users.

In this commit, we work around this limitation:

We first try to open the mountpoint as root, and if we get `EACCES`, we retry as the user who started fusermount, and see if we get `ENOTCONN`.

In my testing, I found that `setfsuid` and `setfsgid` don't work to get around the lack of `allow_other`.  (Sorry, I don't know enough about the Linux kernel to tell whether that's significant.)  As a workaround, I decided to use `setresuid` and `setresgid` in a forked child process, and communicate via its exit status.

Please give feedback on correctness, style and suggest tests.

Fixes https://github.com/libfuse/libfuse/issues/586

### CODE REVIEW

File: util/fusermount.c [hunk 1]
```diff
@@ -1190,6 +1190,40 @@ static int send_fd(int sock_fd, int fd)
 	return 0;
 }
 
+static int mnt_is_ENOTCONN_for_owner(const char *mnt)
```
Comment (bsbernd) [inline util/fusermount.c:1193]: Would you mind to add the explanation in the commit message as function comment? Like

/*
 * fusermount typically has the s-bit set - initial open was as root and got EACCESS
 * as 'allow_other' was not specified.
 * Try again with uid of the calling process process
 */

File: util/fusermount.c [hunk 2]
```diff
@@ -1225,10 +1259,21 @@ static int should_auto_unmount(const char *mnt, const char *type)
...
 	if (fd != -1) {
 		close(fd);
 	} else {
-		result = errno == ENOTCONN;
+		switch(errno) {
+		case ENOTCONN:
+			result = 1;
+			break;
+		case EACCES:
+			result = mnt_is_ENOTCONN_for_owner(mnt);
```
Comment (bsbernd) [inline util/fusermount.c:1271]: I find the function name a bit confusing and only understood it, once I read the code. Maybe
something `?recheck_ENOTCONN_as_owner`

Comment (matthiasgoergens) [reply util/fusermount.c:1271]: Done!

File: util/fusermount.c (hunk 1 above)
Comment (matthiasgoergens) [reply util/fusermount.c]: Done.

**Comment by matthiasgoergens** [comment]
@bsbernd Thanks for the quick review.  I pushed a new version that (hopefully!) addresses your remarks.

**Review by bsbernd — APPROVED** [review]
Thanks for the quick updates! Looks good to me.

**Pull request merged**